x/emissions/module/abci.go

```
@@ -74,42 +74,41 @@ func EndBlocker(ctx context.Context, am AppModule) error {
 	}
 
 	// Close any open windows due this blockHeight
 	workerWindowsToClose := am.keeper.GetWorkerWindowTopicIds(sdkCtx, blockHeight)
 	if len(workerWindowsToClose.TopicIds) > 0 {
 		for _, topicId := range workerWindowsToClose.TopicIds {
 			sdkCtx.Logger().Info("ABCI EndBlocker: Worker close cadence met for topic", "topicId", topicId)
+
 			// Check if there is an unfulfilled nonce
 			nonces, err := am.keeper.GetUnfulfilledWorkerNonces(sdkCtx, topicId)
 			if err != nil {
-				sdkCtx.Logger().Warn("Error getting unfulfilled worker nonces", "error", err)
-				continue
+				return errors.WrapWithFields(err, "failed: get unfulfilled worker nonces for topic", "topic_id", topicId)
 			} else if len(nonces.Nonces) == 0 {
 				// No nonces to fulfill
 				continue
-			} else {
-				topic, err := am.keeper.GetTopic(sdkCtx, topicId)
+			}
+
+			topic, err := am.keeper.GetTopic(sdkCtx, topicId)
+			if err != nil {
+				return errors.WrapWithFields(err, "failed: fetch topic", "topic_id", topicId)
+			}
+
+			for _, nonce := range nonces.Nonces {
+				// No need to validate blockHeight boundaries - we accept submissions until this block.
+				sdkCtx.Logger().Debug("ABCI EndBlocker: closing worker window for topic", "topic_id", topicId, "nonce", nonce)
+				err = allorautils.CloseWorkerNonce(&am.keeper, sdkCtx, topic, *nonce)
 				if err != nil {
-					sdkCtx.Logger().Warn("Error getting topic", "error", err)
-					continue
-				}
-				for _, nonce := range nonces.Nonces {
-					// No need to validate blockHeight boundaries - we accept submissions until this block.
-					sdkCtx.Logger().Debug("ABCI EndBlocker", "blockHeight", blockHeight, "closing worker window for topic", "topicId", topicId, "nonce", nonce)
-					err = allorautils.CloseWorkerNonce(&am.keeper, sdkCtx, topic, *nonce)
-					if err != nil {
-						sdkCtx.Logger().Info("Error closing worker nonce", "error", err)
-					}
+					return errors.WrapWithFields(err, "failed: close worker nonce", "topic_id", topicId)
```

If we want to return the error here, on any case of any error, then we need to carefully review all of the errors we currently have, and either classify them as unexpected (then halt) or recoverable, or just do not return and propagate such recoverable errors on those functions if they are expected to be handled.

This change would need careful review of our error handling within all Begin/EndBlocker code, so it'd be better done in a separate effort than the current fixes on BeginBlocker/EndBlocker.

---

And I think we would need to do that, for safety; because we are treating equally (= ignoring) both a business-as-usual error like "no inferers in this epoch" or a critical one like "no access to store".